### PR TITLE
Add checksum to getdescriptorinfo

### PR DESCRIPTION
**Pull request by sipa**

### CODE REVIEW

**Comment by gwillen** [comment]
So the only reason I didn't write it this way, is that people have said they don't want private keys to be echoed in the output.

Personally I think this is fine -- the only way to get a private key in the output, is if you include it in the input. And in fact, there has to be SOME way to get private keys echoed back, if you want to be able to get checksums for descriptors with private keys included.

So I'd advocate for this PR the way it is, but if people aren't comfortable with it, I'd suggest adding a flag, and not echoing descriptors with private keys unless the flag is set.

Thanks!

(Tagging #15740 for the record.)

**Comment by meshcollider** [comment]
utACK https://github.com/bitcoin/bitcoin/pull/15986/commits/15c482ef711a166b74dcd1fa127d213799d2a1d1

**Comment by Sjors** [comment]
Concept ACK.

Light preference for returning only the checksum rather than echoing a private key. But I do agree with @gwillen's point: 
>the only way to get a private key in the output, is if you include it in the input.

I find the extra  `with_checksum` confusing. Maybe we can add an argument "keep_private_keys" which defaults to `false`. I would also prefer sticking to canonical form (but with private keys). An additional argument `canonical` can be `true` by default. I know it makes the change a bit more complicated, but it seems more robust in the long run.

**Comment by jb55** [comment]
I'm confused by this, getdescriptorinfo already adds a checksum if it's missing?

**Comment by sipa** [comment]
@jb55 It also normalizes the descriptor (which among other things turns WIP into pubkey, and xprv into xpub)

**Comment by jb55** [comment]
ah ok that was the bit I was missing

**Comment by promag** [comment]
ACK 15c482ef711a166b74dcd1fa127d213799d2a1d1, only reviewed code change. Looks like returning only the checksum, like @Sjors suggested, is a simple solution to not echo back private key. A small release note could be added.

Refactor commit best viewed with `git show --color-words 7976dd65f6503082370a40ac189b925322c29918`.

**Comment by achow101** [comment]
Code Review ACK 15c482ef711a166b74dcd1fa127d213799d2a1d1

 I would prefer if there were also a test where the descriptor were slightly modified (e.g. replacing `'` with `h` in origin info) and that the result of `getdescriptorinfo` gives the correct canonical form and the modified form with the checksum.

**Review by achow101 — APPROVED** [review]

**Comment by Sjors** [comment]
Tested ACK 15c482e

**Comment by sipa** [comment]
I like the suggestion of instead just reporting the checksum for the unmodified descriptor, and not echoing anything back. Since this has been reviewed already, I'm happy to go either way. What do reviewers think?

**Comment by achow101** [comment]
> I like the suggestion of instead just reporting the checksum for the unmodified descriptor, and not echoing anything back. Since this has been reviewed already, I'm happy to go either way. What do reviewers think?

I prefer this because it is unambiguous as to what the checksum belongs to and what the descriptor with with checksum will look like.

**Comment by sipa** [comment]
Updated to instead add a `"checksum"` to `getdescriptorinfo` that returns the checksum for the (unmodified) descriptor input.

File: src/rpc/misc.cpp [hunk 1]
```diff
@@ -136,6 +136,7 @@ UniValue getdescriptorinfo(const JSONRPCRequest& request)
             RPCResult{
             "{\n"
             "  \"descriptor\" : \"desc\",         (string) The descriptor in canonical form, without private keys\n"
+            "  \"checksum\" : \"chksum\",         (string) The checksum for the input descriptor\n"
```
Comment (Sjors) [inline src/rpc/misc.cpp]: nit: did you really mean `chksum`?

**Review by Sjors — APPROVED** [review]
ACK 26d3fad1093dfc697048313be7a96c9adf723654

**Comment by achow101** [comment]
Code Review ACK 26d3fad1093dfc697048313be7a96c9adf723654

**Comment by fanquake** [comment]
Two wallet reviewer ACKs here; @meshcollider want to take a final look? I've compiled to check that there aren't any silent merge conflicts etc.

**Comment by meshcollider** [comment]
re-Code Review ACK 26d3fad1093dfc697048313be7a96c9adf723654

**Pull request merged**